src/Nethermind/Nethermind.Merge.Plugin/EngineRpcModule.cs

```diff
@@ -72,14 +76,31 @@ public async Task<ResultWrapper<PayloadStatusV1>> engine_newPayloadV1(ExecutionP
             return await NewPayload(executionPayload, nameof(engine_newPayloadV1));
         }
 
-        public Task<ResultWrapper<PayloadStatusV1>> engine_newPayloadV2(ExecutionPayload executionPayload) =>
-            NewPayload(executionPayload, nameof(engine_newPayloadV2));
+        public Task<ResultWrapper<PayloadStatusV1>> engine_newPayloadV2(ExecutionPayload executionPayload)
+        {
+            var spec = _specProvider.GetSpec((executionPayload.BlockNumber, executionPayload.Timestamp));
+            string? error = null;
+
+            if (spec.WithdrawalsEnabled && executionPayload.Withdrawals is null)
+                error = "ExecutionPayloadV2 expected";
+            else if (!spec.WithdrawalsEnabled && executionPayload.Withdrawals is not null)
+                error = "ExecutionPayloadV1 expected";
```

Comment:
Would like to see versioning and validation abstracted at least to the extension methods, to be clearer.